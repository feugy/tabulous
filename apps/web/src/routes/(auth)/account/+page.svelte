<script>
  import {
    Button,
    Header,
    Input,
    PlayerThumbnail,
    Progress
  } from '@src/components'
  import { updateCurrentPlayer } from '@src/stores'
  import { translateError } from '@src/utils'
  import { debounceTime, finalize, from, Subject, switchMap, tap } from 'rxjs'
  import { onDestroy } from 'svelte'
  import { _ } from 'svelte-intl'

  import { invalidateAll } from '$app/navigation'

  import AvatarDialogue from './AvatarDialogue.svelte'

  /** @type {import('./$types').PageData} */
  export let data = {}

  const username$ = new Subject()
  let user = data.session.player
  let isSaving = false
  let usernameError = null
  let openAvatarDialogue = false
  let avatar = user.avatar

  const saveSubscription = username$
    .pipe(
      tap(() => (usernameError = null)),
      debounceTime(500),
      switchMap(username => {
        isSaving = true
        return from(
          updateCurrentPlayer(username)
            // updates page data with new user details
            .then(invalidateAll)
            .then(() => (user.username = username))
            .catch(error => (usernameError = error))
        ).pipe(finalize(() => (isSaving = false)))
      })
    )
    .subscribe()

  onDestroy(() => saveSubscription.unsubscribe())

  function handleSave({ target }) {
    username$.next(target.value)
  }

  async function handleCloseAvatarDialogue({ detail: confirmed }) {
    if (confirmed) {
      user.avatar = (await updateCurrentPlayer(user.username, avatar)).avatar
    }
    avatar = user.avatar
  }
</script>

<svelte:head>
  <title>{$_('page-titles.account')}</title>
</svelte:head>

<main>
  <Header
    {user}
    breadcrumb={[
      { label: $_('labels.home'), href: '/home' },
      { label: $_('labels.account') }
    ]}
  >
    <h1>
      {$_('titles.account')}
    </h1>
  </Header>
  <section>
    <h2>{$_('titles.auth-provider')}</h2>
    <fieldset>
      {#if user.provider}
        <div>{$_('labels.oauth-provider', user)}</div>
      {:else}
        <div>{$_('labels.manual-provider', user)}</div>
      {/if}
      <label for="username">{$_('labels.username')}</label>
      <span class="with-progress">
        <Input
          name="username"
          value={user.username}
          disabled={isSaving}
          on:input={handleSave}
        />
        <span
          >{#if isSaving}<Progress radius={24} />{/if}</span
        >
        {#if usernameError}
          <span class="error">{translateError($_, usernameError)}</span>
        {/if}
      </span>
      <span>{$_('labels.avatar')}</span>
      <span class="aligned">
        <PlayerThumbnail player={user} dimension={150} />
        <Button
          text={$_('actions.change-avatar')}
          on:click={() => (openAvatarDialogue = true)}
        />
        <AvatarDialogue
          bind:open={openAvatarDialogue}
          bind:avatar
          on:close={handleCloseAvatarDialogue}
        />
      </span>
    </fieldset>
  </section>
</main>

<style lang="postcss">
  main {
    @apply flex flex-col h-full w-full;
  }

  h1 {
    @apply text-3xl py-4;
  }

  h2 {
    @apply text-2xl pb-4;
  }

  section {
    @apply my-8 p-8 self-center lg:w-3/4 border rounded-md;
  }

  fieldset {
    @apply grid gap-x-8 gap-y-4 grid-cols-[auto,1fr] items-center;

    div {
      @apply col-span-2;
    }
  }

  .with-progress {
    @apply grid grid-cols-[1fr,auto] items-center;
  }

  .aligned {
    @apply flex items-center gap-4;
  }

  .error {
    @apply text-$accent-warm;
  }
</style>
