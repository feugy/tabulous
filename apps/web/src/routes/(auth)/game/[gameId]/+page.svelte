<script>
  import { GameAside, InvitePlayerDialogue, Progress } from '@src/components'
  import {
    actionMenuProps,
    cameraSaves,
    connected,
    currentCamera,
    currentGame,
    gamePlayerById,
    handMeshes,
    handVisible,
    highlightHand,
    initEngine,
    initIndicators,
    joinGame,
    leaveGame,
    longInputs,
    meshDetails,
    playerColor,
    restoreCamera,
    saveCamera,
    sendToThread,
    thread,
    toastError,
    toastInfo,
    visibleFeedbacks,
    visibleIndicators
  } from '@src/stores'
  import { isLobby, observeDimension } from '@src/utils'
  import { onMount } from 'svelte'
  import { _ } from 'svelte-intl'

  import { beforeNavigate, goto } from '$app/navigation'
  import { page } from '$app/stores'

  import CameraSwitch from './CameraSwitch.svelte'
  import CursorInfo from './CursorInfo.svelte'
  import GameHand from './GameHand.svelte'
  import GameMenu from './GameMenu.svelte'
  import Indicators from './Indicators.svelte'
  import MeshDetails from './MeshDetails.svelte'
  import Parameters from './Parameters'
  import RadialMenu from './RadialMenu.svelte'

  /** @type {import('./$types').PageData} */
  export let data = {}

  const longTapDelay = 250
  let engine
  let canvas
  let interaction
  let hand
  let openInviteDialogue = false
  let joinPromise
  let dimensionObserver
  let dimensionSubscription
  let gameParameters

  onMount(async () => {
    engine = initEngine({ canvas, interaction, longTapDelay, hand })
    initIndicators({ engine, canvas, hand })
    askForGame()
    dimensionObserver = observeDimension(interaction, 0)
    dimensionSubscription = dimensionObserver.dimension$.subscribe(
      ({ width, height }) => {
        canvas.style.width = `${width}px`
        canvas.style.height = `${height}px`
        engine?.resize()
      }
    )
  })

  beforeNavigate(() => {
    leaveGame(data.session.player)
    engine?.dispose()
    dimensionObserver?.disconnect()
    dimensionSubscription?.unsubscribe()
  })

  function askForGame(parameters) {
    gameParameters = null
    joinPromise = joinGame({
      gameId: $page.params.gameId,
      ...data.session,
      parameters,
      onDeletion: () => {
        toastInfo({ contentKey: 'labels.game-deleted-by-owner' })
        goto('/home')
      }
    })
      .then(result => {
        if (result?.schemaString) {
          gameParameters = {
            schema: JSON.parse(result.schemaString)
          }
        }
        if (isLobby(result)) {
          goto('/home')
        }
      })
      .catch(err => {
        console.error(err)
        toastError({ content: err.message })
        goto('/home')
      })
  }

  function handleCloseDetails() {
    interaction?.focus()
  }

  function handleSubmitParameters({ detail: parameters }) {
    askForGame(parameters)
  }
</script>

<svelte:head>
  <title>{$_('page-titles.game')}</title>
</svelte:head>

{#await joinPromise}
  <div class="overlay">
    <Progress />
  </div>
{/await}
<aside class="top">
  <GameMenu on:invite-player={() => (openInviteDialogue = true)} />
  <CameraSwitch
    {longTapDelay}
    current={$currentCamera}
    saves={$cameraSaves}
    on:longTap={() => longInputs.next()}
    on:restore={({ detail: { index } }) => restoreCamera(index)}
    on:save={({ detail: { index } }) => saveCamera(index)}
  />
  <InvitePlayerDialogue
    game={$currentGame}
    open={openInviteDialogue}
    on:close={() => (openInviteDialogue = false)}
  />
</aside>
<main>
  <!-- svelte-ignore a11y-autofocus -->
  <div
    class="interaction"
    tabindex="0"
    role="textbox"
    autofocus
    bind:this={interaction}
    on:contextmenu|preventDefault
  >
    <canvas bind:this={canvas} />
    <Indicators items={$visibleIndicators} feedbacks={$visibleFeedbacks} />
    <GameHand
      visible={$handVisible}
      highlight={$highlightHand}
      meshes={$handMeshes}
      color={$playerColor}
      bind:node={hand}
    />
    {#if gameParameters}
      <Parameters
        schema={gameParameters.schema}
        on:submit={handleSubmitParameters}
      />
    {/if}
    <MeshDetails mesh={$meshDetails} on:close={handleCloseDetails} />
  </div>
  <RadialMenu {...$actionMenuProps || {}} />
  <CursorInfo halos={longInputs} />
  <GameAside
    game={$currentGame}
    player={data.session.player}
    playerById={$gamePlayerById}
    connected={$connected}
    thread={$thread}
    on:sendMessage={({ detail }) => sendToThread(detail.text)}
  />
</main>

<style lang="postcss">
  main,
  .overlay {
    @apply absolute inset-0 flex items-stretch overflow-hidden;
  }

  main {
    container-type: inline-size;
  }

  .interaction {
    @apply select-none flex-1 relative;
    touch-action: none;
  }

  canvas {
    @apply absolute top-0 left-0 select-none h-full w-full;
    touch-action: none;
  }

  aside.top {
    @apply absolute z-10 top-0 left-0 p-2 flex gap-2;
  }

  .overlay {
    @apply flex items-center justify-center z-10;
  }
</style>
