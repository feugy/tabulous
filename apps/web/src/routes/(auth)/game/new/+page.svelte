<script>
  import { Button, Header } from '@src/components'
  import { createGame } from '@src/stores'
  import { translateError } from '@src/utils'
  import { onMount } from 'svelte'
  import { get } from 'svelte/store'
  import { _ } from 'svelte-intl'

  import { goto } from '$app/navigation'
  import { page } from '$app/stores'

  /** @type {import('./$types').PageData} */
  export let data = {}

  const gameName = $page.url.searchParams.get('name')
  let error

  onMount(async () => {
    if (gameName) {
      try {
        goto(await createGame(gameName), { replaceState: true })
      } catch (err) {
        error = translateError(get(_), err)
      }
    } else {
      goto('/home')
    }
  })
</script>

<svelte:head>
  <title>{$_('page-titles.create-game')}</title>
</svelte:head>

<main>
  <Header
    user={data.session?.player}
    breadcrumb={[
      { label: $_('labels.home'), href: '/' },
      { label: $_('titles.create-game') }
    ]}
  />
  <section>
    {#if error}
      <p>{error}</p>
    {/if}
    <a href="/home"><Button text={$_('labels.home')} icon="arrow_back" /></a>
  </section>
</main>

<style lang="postcss">
  main {
    @apply flex flex-col w-full pb-8;
  }

  section {
    @apply flex flex-col lg:w-3/4 lg:mx-auto;
  }

  p {
    @apply text-2xl pt-8 pb-4;
  }
</style>
