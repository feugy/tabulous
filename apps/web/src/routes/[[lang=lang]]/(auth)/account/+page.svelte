<script>
  // @ts-check
  /** @typedef {import('@src/graphql').FullPlayer} Player */

  import {
    Button,
    Header,
    Input,
    PageFooter,
    Pane,
    PlayerThumbnail,
    Progress,
    UsernameSearchability
  } from '@src/components'
  import { updateCurrentPlayer } from '@src/stores'
  import { translateError } from '@src/utils'
  import { debounceTime, finalize, from, Subject, switchMap, tap } from 'rxjs'
  import { onDestroy } from 'svelte'
  import { _ } from 'svelte-intl'

  import { invalidate } from '$app/navigation'

  import AvatarDialogue from './AvatarDialogue.svelte'

  /** @type {import('./$types').PageData} */
  export let data

  /** @type {Subject<string>} */
  const username$ = new Subject()
  /** @type {Player} */
  let user = data.session?.player ?? {
    username: '',
    id: '',
    currentGameId: null
  }
  let isSaving = false
  /** @type {?Error} */
  let usernameError = null
  let openAvatarDialogue = false
  let avatar = user.avatar

  const saveSubscription = username$
    .pipe(
      tap(() => (usernameError = null)),
      debounceTime(500),
      switchMap(username => {
        isSaving = true
        return from(
          updateCurrentPlayer(username)
            // updates page data with new user details
            .then(() => invalidate('data:session'))
            .then(() => (user.username = username))
            .catch(error => (usernameError = error))
        ).pipe(finalize(() => (isSaving = false)))
      })
    )
    .subscribe()

  onDestroy(() => saveSubscription.unsubscribe())

  function handleSave(/** @type {Event} */ { target }) {
    username$.next(/** @type {HTMLInputElement} */ (target).value)
  }

  async function handleCloseAvatarDialogue(
    /** @type {CustomEvent<boolean>} */ { detail: confirmed }
  ) {
    if (confirmed) {
      user.avatar = (await updateCurrentPlayer(user.username, avatar)).avatar
    }
    avatar = user.avatar
  }
</script>

<svelte:head>
  <title>{$_('page-titles.account')}</title>
</svelte:head>

<main>
  <i id="top" />
  <Header {user}>
    <h1>
      {$_('titles.account')}
    </h1>
  </Header>
  <section>
    <Pane title={$_('titles.auth-provider')}>
      <fieldset>
        {#if user.provider}
          <div>{$_('labels.oauth-provider', user)}</div>
        {:else}
          <div>{$_('labels.manual-provider', user)}</div>
        {/if}
        <label for="username">{$_('labels.username')}</label>
        <span class="with-progress">
          <Input
            name="username"
            value={user.username}
            disabled={isSaving}
            on:input={handleSave}
          />
          <span
            >{#if isSaving}<Progress radius={24} />{/if}</span
          >
          {#if usernameError}
            <span class="error">{translateError($_, usernameError)}</span>
          {/if}
        </span>
        <label for="avatar">{$_('labels.avatar')}</label>
        <span class="aligned">
          <PlayerThumbnail player={user} dimension={150} />
          <Button
            id="avatar"
            text={$_('actions.change-avatar')}
            on:click={() => (openAvatarDialogue = true)}
          />
          <AvatarDialogue
            bind:open={openAvatarDialogue}
            bind:avatar
            on:close={handleCloseAvatarDialogue}
          />
        </span>
        <span>{$_('labels.searchability')}</span>
        <UsernameSearchability searchable={user.usernameSearchable} />
      </fieldset>
    </Pane>
  </section>
  <PageFooter />
</main>

<style lang="postcss">
  main {
    @apply flex flex-col;
  }

  section {
    @apply flex flex-col flex-1 p-6 self-center w-full xl:w-screen-xl;
  }

  fieldset {
    @apply mt-4 grid gap-x-8 gap-y-4 grid-cols-[auto,1fr] items-center;

    :nth-child(even) {
      font-family: var(--font-heading);
    }

    div {
      @apply col-span-2;
    }
  }

  .with-progress {
    @apply grid grid-cols-[1fr,auto] items-center;
  }

  .aligned {
    @apply flex items-center gap-8 flex-wrap;
  }

  .error {
    @apply text-$accent-warm;
  }
</style>
