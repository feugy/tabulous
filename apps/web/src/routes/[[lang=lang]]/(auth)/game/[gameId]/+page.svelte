<script>
  import { _ } from 'svelte-intl'

  import Loading from './LoadingScreen'

  /** @type {import('./$types').PageData} */
  export let data = {}
</script>

<svelte:head>
  <title>{$_('page-titles.game')}</title>
</svelte:head>
{#await import('./Page.svelte')}
  <Loading />
{:then { default: page }}
  <svelte:component this={page} {data} />
{/await}
