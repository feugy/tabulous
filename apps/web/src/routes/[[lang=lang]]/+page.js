import { redirect } from '@sveltejs/kit'

/** @type {import('./$types').PageLoad} */
export function load({ params: { lang } }) {
  throw redirect(307, lang ? `/${lang}/home` : '/home')
}
