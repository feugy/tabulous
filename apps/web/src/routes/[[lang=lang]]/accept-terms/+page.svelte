<script>
  import { Header, PageFooter } from '@src/components'
  import { _ } from 'svelte-intl'

  import { page } from '$app/stores'

  import Form from './Form.svelte'
  import ScrollableTerms from './ScrollableTerms.svelte'

  /** @type {import('./$types').PageData} */
  export let data = {}

  const redirect = $page.url?.searchParams.get('redirect')

  let disabled = true
</script>

<svelte:head>
  <title>{$_('page-titles.accept-terms')}</title>
</svelte:head>

<main>
  <Header user={data.session?.player}><h1>{$_('titles.legal')}</h1></Header>
  <section>
    <h2>{$_('titles.welcome')}</h2>
    <p>{$_('labels.terms-intro')}</p>
    <ScrollableTerms on:end={() => (disabled = false)} />
    <Form {disabled} {redirect} />
  </section>
  <PageFooter />
</main>

<style lang="postcss">
  main {
    @apply flex flex-col max-h-screen overflow-auto;
  }

  section {
    @apply flex flex-col flex-1 px-6 self-center w-full xl:w-screen-xl overflow-hidden;

    & > p {
      @apply mb-6;
    }
  }
</style>
