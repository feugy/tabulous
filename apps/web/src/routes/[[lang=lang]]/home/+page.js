// @ts-check
import { createGame, listCatalog, listGames } from '@src/stores'
import { redirect } from '@sveltejs/kit'

/** @type {import('./$types').PageLoad} */
export async function load({ parent, url, params: { lang } }) {
  const data = await parent()
  const hasPlayer = data.session?.player
  const [catalog, currentGames] = await Promise.all([
    listCatalog(),
    hasPlayer ? listGames() : Promise.resolve(null)
  ])
  /** @type {?Error} */
  let creationError = null
  const name = url.searchParams.get('game-name')
  if (name && hasPlayer) {
    /** @type {?import('@src/graphql').Game['id']} */
    let gameId = null
    try {
      gameId = (await createGame(name)).id
    } catch (err) {
      creationError = /** @type {Error} */ (err)
    }
    if (gameId) {
      throw redirect(307, `${lang ? `/${lang}` : ''}/game/${gameId}`)
    }
  }
  return { catalog, currentGames, creationError }
}
