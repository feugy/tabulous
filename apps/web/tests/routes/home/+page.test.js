import { faker } from '@faker-js/faker'
import { load } from '@src/routes/home/+page'
import HomePage from '@src/routes/home/+page.svelte'
import {
  createGame,
  deleteGame,
  listCatalog,
  listFriends,
  listGames,
  toastError,
  toastInfo
} from '@src/stores'
import { fireEvent, render, screen, within } from '@testing-library/svelte'
import { translate } from '@tests/test-utils'
import { readable } from 'svelte/store'
import html from 'svelte-htm'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import { goto } from '$app/navigation'

vi.mock('@src/stores', async () => {
  const stores = await vi.importActual('@src/stores')
  return {
    ...stores,
    createGame: vi.fn(),
    deleteGame: vi.fn(),
    listCatalog: vi.fn(),
    listGames: vi.fn(),
    listFriends: vi.fn(),
    receiveGameListUpdates: vi.fn(),
    toastError: vi.fn(),
    toastInfo: vi.fn()
  }
})

beforeEach(vi.clearAllMocks)

describe('/home route loader', () => {
  it('loads catalog only for anonymous user', async () => {
    const parent = async () => ({ session: null })
    const catalog = [{ id: 'game-1' }, { id: 'game-2' }]
    listCatalog.mockResolvedValueOnce(catalog)
    expect(
      await load({ parent, url: new URL('https://tabulous.fr/home') })
    ).toEqual({
      catalog,
      creationError: null,
      currentGames: null
    })
    expect(listCatalog).toHaveBeenCalledTimes(1)
    expect(listGames).not.toHaveBeenCalled()
  })

  it('loads catalog and current games for authenticated user', async () => {
    const parent = async () => ({ session: { player: { name: 'dude' } } })
    const catalog = [{ id: 'game-1' }, { id: 'game-2' }]
    const currentGames = [{ id: 'game-3' }]
    listCatalog.mockResolvedValueOnce(catalog)
    listGames.mockResolvedValueOnce(currentGames)
    expect(
      await load({ parent, url: new URL('https://tabulous.fr/home') })
    ).toEqual({
      catalog,
      creationError: null,
      currentGames
    })
    expect(listCatalog).toHaveBeenCalledTimes(1)
    expect(listGames).toHaveBeenCalledTimes(1)
  })

  it('does not create game when unauthenticated', async () => {
    const parent = async () => ({ session: null })
    const catalog = [
      { id: 'game-1', name: 'klondike' },
      { id: 'game-2', name: 'chess' }
    ]
    listCatalog.mockResolvedValueOnce(catalog)
    expect(
      await load({
        parent,
        url: new URL(`https://tabulous.fr/home?game-name=${catalog[0].name}`)
      })
    ).toEqual({
      catalog,
      creationError: null,
      currentGames: null
    })
    expect(listCatalog).toHaveBeenCalledTimes(1)
    expect(listGames).not.toHaveBeenCalled()
    expect(createGame).not.toHaveBeenCalled()
  })

  describe('given an authenticated user', () => {
    const parent = async () => ({ session: { player: { name: 'dude' } } })
    const catalog = [
      { id: 'game-1', name: 'klondike' },
      { id: 'game-2', name: 'chess' }
    ]
    const currentGames = [{ id: 'game-3' }]

    beforeEach(() => {
      listCatalog.mockResolvedValueOnce(catalog)
      listGames.mockResolvedValueOnce(currentGames)
    })

    it('redirects to new game on creation', async () => {
      const id = faker.string.uuid()
      createGame.mockResolvedValueOnce({ id })
      await expect(
        load({
          parent,
          url: new URL(`https://tabulous.fr/home?game-name=${catalog[1].name}`)
        })
      ).rejects.toEqual({ status: 307, location: `/game/${id}` })
      expect(listCatalog).toHaveBeenCalledTimes(1)
      expect(listGames).toHaveBeenCalledTimes(1)
      expect(createGame).toHaveBeenCalledWith(catalog[1].name)
      expect(createGame).toHaveBeenCalledTimes(1)
    })

    it('returns game creation error', async () => {
      const error = new Error(`You own 6 games, you can not create more`)
      createGame.mockRejectedValueOnce(error)
      expect(
        await load({
          parent,
          url: new URL(`https://tabulous.fr/home?game-name=${catalog[1].name}`)
        })
      ).toEqual({ creationError: error, currentGames, catalog })
      expect(listCatalog).toHaveBeenCalledTimes(1)
      expect(listGames).toHaveBeenCalledTimes(1)
      expect(createGame).toHaveBeenCalledWith(catalog[1].name)
      expect(createGame).toHaveBeenCalledTimes(1)
    })
  })
})

describe('/home route', () => {
  const parent = async () => ({ session: { player: { name: 'dude' } } })
  const catalog = [
    { id: 'game-1', name: 'klondike', locales: { fr: { title: 'Solitaire' } } },
    { id: 'game-2', name: 'draughts', locales: { fr: { title: 'Dames' } } }
  ]
  const games = [
    {
      id: 'game-3',
      kind: 'chess',
      locales: { fr: { title: 'Echecs' } },
      players: []
    }
  ]

  beforeEach(async () => {
    listCatalog.mockResolvedValueOnce(catalog)
    listGames.mockResolvedValueOnce(games)
    listFriends.mockReturnValueOnce(readable([]))
  })

  async function renderWithLoad(gameName) {
    const data = await load({
      parent,
      url: new URL(
        `https://tabulous.fr/home${gameName ? `?game-name=${gameName}` : ''}`
      )
    })
    return render(
      html`<${HomePage} data=${{ ...data, ...(await parent()) }} />`
    )
  }

  it('browse to game creation', async () => {
    await renderWithLoad()
    fireEvent.click(
      screen.getByRole('heading', { name: catalog[1].locales.fr.title })
    )
    expect(goto).toHaveBeenCalledWith(`/home?game-name=${catalog[1].name}`)
    expect(goto).toHaveBeenCalledTimes(1)
  })

  it('displays error when too many games where created', async () => {
    const count = faker.number.int({ min: 4, max: 10 })
    createGame.mockRejectedValueOnce(
      new Error(`You own ${count} games, you can not create more`)
    )
    await renderWithLoad(catalog[0].name)

    expect(toastError).toHaveBeenCalledWith({
      content: translate('errors.too-many-games', { count })
    })
    expect(toastError).toHaveBeenCalledTimes(1)
    expect(goto).not.toHaveBeenCalled()
  })

  it('displays error game is restricted', async () => {
    createGame.mockRejectedValueOnce(new Error('Access to game is restricted'))
    await renderWithLoad(catalog[0].name)

    expect(toastError).toHaveBeenCalledWith({
      content: translate('errors.restricted-game')
    })
    expect(toastError).toHaveBeenCalledTimes(1)
    expect(goto).not.toHaveBeenCalled()
  })

  it('can cancels game deletion', async () => {
    await renderWithLoad()
    const gameLink = screen
      .getByRole('heading', {
        name: games[0].locales.fr.title
      })
      .closest('article')
    await fireEvent.click(
      within(gameLink).getByRole('button', { name: 'delete' })
    )

    const confirmDialogue = screen.getByRole('dialog')
    expect(confirmDialogue).toBeInTheDocument()

    await fireEvent.click(within(confirmDialogue).getAllByRole('button')[0])

    expect(deleteGame).not.toHaveBeenCalled()
    expect(toastInfo).not.toHaveBeenCalled()
  })

  it('displays a toaster on game deletion', async () => {
    await renderWithLoad()
    const gameLink = screen
      .getByRole('heading', {
        name: games[0].locales.fr.title
      })
      .closest('article')
    await fireEvent.click(
      within(gameLink).getByRole('button', { name: 'delete' })
    )

    const confirmDialogue = screen.getByRole('dialog')
    expect(confirmDialogue).toBeInTheDocument()

    await fireEvent.click(within(confirmDialogue).getAllByRole('button')[1])

    expect(toastInfo).toHaveBeenCalledWith({
      contentKey: 'labels.game-deleted'
    })
    expect(toastInfo).toHaveBeenCalledTimes(1)
    expect(deleteGame).toHaveBeenCalledWith(games[0].id)
    expect(deleteGame).toHaveBeenCalledTimes(1)
  })
})
