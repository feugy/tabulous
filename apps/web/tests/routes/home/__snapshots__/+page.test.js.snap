// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/home route displays anonymous catalog 1`] = `
<body>
  <div>
     
    <main>
      <header>
        <nav>
          <button
            class="secondary"
          >
            <span
              class="material-icons"
            >
              login
            </span>
            
            
            
          </button>
        </nav>
         
        <div>
          <h1>
            Bienvenue sur Tabulous !
          </h1>
        </div>
      </header>
       
      <div>
         
        <h2>
          Jeux disponibles
        </h2>
         
        <section
          aria-roledescription="catalog"
        >
          <article>
            <img
              alt="Jeu de 32 cartes"
              src="http://localhost:3001/games/32-cards/catalog/cover.webp"
            />
             
            <div
              class="content"
            >
              <legend>
                <div
                  class="title"
                >
                  <h3>
                    Jeu de 32 cartes
                  </h3>
                   
                  <details
                    class="hidden"
                  >
                    <summary>
                      <strong>
                        Créé par : 
                      </strong>
                      undefined
                    </summary>
                     
                    <p>
                      <strong>
                        Illustré par : 
                      </strong>
                      undefined
                    </p>
                     
                    <p>
                      <strong>
                        Distribué par : 
                      </strong>
                      undefined
                    </p>
                  </details>
                </div>
                 
                <span
                  class="characteristics"
                >
                  <span
                    class="hidden"
                  >
                    <span
                      class="material-icons"
                    >
                      people
                    </span>
                    undefined
                  </span>
                   
                  <span
                    class="hidden"
                  >
                    <span
                      class="material-icons"
                    >
                      access_time
                    </span>
                    null'
                  </span>
                   
                  <span
                    class="hidden"
                  >
                    <span
                      class="material-icons"
                    >
                      person
                    </span>
                    null+
                  </span>
                </span>
              </legend>
            </div>
          </article>
          <article>
            <img
              alt="Solitaire"
              src="http://localhost:3001/games/klondike/catalog/cover.webp"
            />
             
            <div
              class="content"
            >
              <legend>
                <div
                  class="title"
                >
                  <h3>
                    Solitaire
                  </h3>
                   
                  <details
                    class="hidden"
                  >
                    <summary>
                      <strong>
                        Créé par : 
                      </strong>
                      undefined
                    </summary>
                     
                    <p>
                      <strong>
                        Illustré par : 
                      </strong>
                      undefined
                    </p>
                     
                    <p>
                      <strong>
                        Distribué par : 
                      </strong>
                      undefined
                    </p>
                  </details>
                </div>
                 
                <span
                  class="characteristics"
                >
                  <span
                    class="hidden"
                  >
                    <span
                      class="material-icons"
                    >
                      people
                    </span>
                    undefined
                  </span>
                   
                  <span>
                    <span
                      class="material-icons"
                    >
                      access_time
                    </span>
                    15'
                  </span>
                   
                  <span>
                    <span
                      class="material-icons"
                    >
                      person
                    </span>
                    7+
                  </span>
                </span>
              </legend>
            </div>
          </article>
        </section>
      </div>
    </main>
     
    
  </div>
</body>
`;

exports[`/home route displays connected catalog 1`] = `
<body>
  <div>
     
    <main>
      <header>
        <nav>
          <button
            class="secondary"
          >
            <span
              class="material-icons"
            >
              directions_run
            </span>
            
            
            
          </button>
        </nav>
         
        <div>
          <h1>
            Bonjour John Doo !
          </h1>
        </div>
      </header>
       
      <div>
        <h2>
          Parties en cours
        </h2>
         
        <section
          aria-roledescription="games"
        >
          
          <article>
            <span
              class="title"
            >
              <h3>
                Jeu de 32 cartes
              </h3>
               
              <button
                class="secondary"
              >
                <span
                  class="material-icons"
                >
                  delete
                </span>
                
                
                
              </button>
            </span>
             
            <span
              class="created"
            >
              15 juil., 10:15
            </span>
             
            <span
              class="players"
            >
              avec 
            </span>
          </article>
        </section>
         
        <h2>
          Jeux disponibles
        </h2>
         
        <section
          aria-roledescription="catalog"
        >
          <article>
            <img
              alt="Jeu de 32 cartes"
              src="http://localhost:3001/games/32-cards/catalog/cover.webp"
            />
             
            <div
              class="content"
            >
              <legend>
                <div
                  class="title"
                >
                  <h3>
                    Jeu de 32 cartes
                  </h3>
                   
                  <details
                    class="hidden"
                  >
                    <summary>
                      <strong>
                        Créé par : 
                      </strong>
                      undefined
                    </summary>
                     
                    <p>
                      <strong>
                        Illustré par : 
                      </strong>
                      undefined
                    </p>
                     
                    <p>
                      <strong>
                        Distribué par : 
                      </strong>
                      undefined
                    </p>
                  </details>
                </div>
                 
                <span
                  class="characteristics"
                >
                  <span
                    class="hidden"
                  >
                    <span
                      class="material-icons"
                    >
                      people
                    </span>
                    undefined
                  </span>
                   
                  <span
                    class="hidden"
                  >
                    <span
                      class="material-icons"
                    >
                      access_time
                    </span>
                    null'
                  </span>
                   
                  <span
                    class="hidden"
                  >
                    <span
                      class="material-icons"
                    >
                      person
                    </span>
                    null+
                  </span>
                </span>
              </legend>
            </div>
          </article>
          <article>
            <img
              alt="Solitaire"
              src="http://localhost:3001/games/klondike/catalog/cover.webp"
            />
             
            <div
              class="content"
            >
              <legend>
                <div
                  class="title"
                >
                  <h3>
                    Solitaire
                  </h3>
                   
                  <details
                    class="hidden"
                  >
                    <summary>
                      <strong>
                        Créé par : 
                      </strong>
                      undefined
                    </summary>
                     
                    <p>
                      <strong>
                        Illustré par : 
                      </strong>
                      undefined
                    </p>
                     
                    <p>
                      <strong>
                        Distribué par : 
                      </strong>
                      undefined
                    </p>
                  </details>
                </div>
                 
                <span
                  class="characteristics"
                >
                  <span
                    class="hidden"
                  >
                    <span
                      class="material-icons"
                    >
                      people
                    </span>
                    undefined
                  </span>
                   
                  <span>
                    <span
                      class="material-icons"
                    >
                      access_time
                    </span>
                    15'
                  </span>
                   
                  <span>
                    <span
                      class="material-icons"
                    >
                      person
                    </span>
                    7+
                  </span>
                </span>
              </legend>
            </div>
          </article>
          <article>
            <img
              alt="6 qui prend"
              src="http://localhost:3001/games/6-takes/catalog/cover.webp"
            />
             
            <div
              class="content"
            >
              <legend>
                <div
                  class="title"
                >
                  <h3>
                    6 qui prend
                  </h3>
                   
                  <details
                    class="hidden"
                  >
                    <summary>
                      <strong>
                        Créé par : 
                      </strong>
                      undefined
                    </summary>
                     
                    <p>
                      <strong>
                        Illustré par : 
                      </strong>
                      undefined
                    </p>
                     
                    <p>
                      <strong>
                        Distribué par : 
                      </strong>
                      undefined
                    </p>
                  </details>
                </div>
                 
                <span
                  class="characteristics"
                >
                  <span
                    class="hidden"
                  >
                    <span
                      class="material-icons"
                    >
                      people
                    </span>
                    undefined
                  </span>
                   
                  <span>
                    <span
                      class="material-icons"
                    >
                      access_time
                    </span>
                    15'
                  </span>
                   
                  <span>
                    <span
                      class="material-icons"
                    >
                      person
                    </span>
                    10+
                  </span>
                </span>
              </legend>
            </div>
          </article>
        </section>
      </div>
    </main>
     
    
  </div>
</body>
`;
